superset-frontend/src/views/CRUD/data/database/DatabaseModal/index.tsx

```diff
@@ -539,32 +517,37 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
       if (validationErrors && !isEmpty(validationErrors)) {
         return;
       }
+      const parameters_schema = isEditMode
+        ? dbToUpdate.parameters_schema.properties
+        : dbModel?.parameters.properties;
+      const additionalEncryptedExtra = JSON.parse(
+        dbToUpdate.encrypted_extra || '{}',
+      );
+      const paramConfigArray = Object.keys(parameters_schema || {});
 
-      const engine = dbToUpdate.backend || dbToUpdate.engine;
-      if (engine === 'bigquery' && dbToUpdate.parameters?.credentials_info) {
-        // wrap encrypted_extra in credentials_info only for BigQuery
+      paramConfigArray.forEach(paramConfig => {
+        // we are going through the parameter configuration and seeing if this is an encrypted field
         if (
-          dbToUpdate.parameters?.credentials_info &&
-          typeof dbToUpdate.parameters?.credentials_info === 'object' &&
-          dbToUpdate.parameters?.credentials_info.constructor === Object
+          parameters_schema[paramConfig]['x-encrypted-extra'] &&
+          dbToUpdate.parameters?.[paramConfig]
         ) {
-          // Don't cast if object
-          dbToUpdate.encrypted_extra = JSON.stringify({
-            credentials_info: dbToUpdate.parameters?.credentials_info,
-          });
-
-          // Convert credentials info string before updating
-          dbToUpdate.parameters.credentials_info = JSON.stringify(
-            dbToUpdate.parameters.credentials_info,
-          );
-        } else {
-          dbToUpdate.encrypted_extra = JSON.stringify({
-            credentials_info: JSON.parse(
-              dbToUpdate.parameters?.credentials_info,
-            ),
-          });
+          if (typeof dbToUpdate.parameters?.[paramConfig] === 'object') {
+            // add new encrypted extra to encrypted_extra object
+            additionalEncryptedExtra[paramConfig] =
+              dbToUpdate.parameters?.[paramConfig];
+            // cast the encrypted field as a string in parameters
```

Comment:
Similarly here and in line 549: it's clear that we're casting it to a string, so the comment doesn't add much. It would be better to say something like:

```js
// The backend expects `encrypted_extra` as a string for historical reasons.
```